packages/react/src/auto/hooks/useRelatedModelOptions.tsx

```diff
@@ -0,0 +1,246 @@
+import { useFieldMetadata } from "./useFieldMetadata.js";
+import { useRelatedModelManager } from "./useRelatedModelManager.js";
+
+export const optionRecordsToLoadCount = 25;
+export const selectedRecordsToLoadCount = 25;
+
+export const useRelatedModelOptions = (props: {
+  field: string; // Field API identifier
+  optionLabel?: OptionLabel; // The label to display for each related model record
+}) => {
+  const { field, optionLabel } = props;
+  const { metadata } = useFieldMetadata(field);
+  const isBelongsToField = metadata.configuration.fieldType === FieldType.BelongsTo;
+
+  // eslint-disable-next-line
+  const { selected } = isBelongsToField ? useLinkedChildModelRelatedModelRecords(props) : useLinkedParentModelRelatedModelRecords(props);
+
+  const { relatedModel, pagination, search } = useAllRelatedModelRecords(props);
+
+  const getOptions = () => {
+    const options = uniqByProperty(
+      [
+        ...getRecordsAsOptions(selected.records ?? [], optionLabel), // Selected records
+        ...getRecordsAsOptions(relatedModel.records, optionLabel), // All related model records
+      ],
+      "id"
+    );
+
+    return options; // TODO - Filter these by search value because the DB selected values are still be shown
+  };
+
+  const [options, setOptions] = useState<any[]>(getOptions());
+
+  useEffect(() => {
+    if (selected.fetching || relatedModel.fetching) {
+      return;
+    }
+
+    setOptions(getOptions());
+  }, [selected.fetching, relatedModel.fetching, getRecordIdsAsString(selected.records), getRecordIdsAsString(relatedModel.records)]);
+
+  return {
+    options,
+
+    selected,
+    relatedModel,
+    pagination,
+    search,
+  };
+};
+
+const getRecordIdsAsString = (records?: { map: (mapperFunction: (record: { id: string }) => string) => string[] }) =>
+  records
+    ?.map((record) => record.id)
+    .sort()
+    .join(",");
+
+export const getRecordsAsOptions = (records: Record<string, any>[], optionLabel?: OptionLabel) => {
+  const getRecordLabel = (record: Record<string, any>, optionLabel?: OptionLabel): string =>
+    optionLabel
+      ? typeof optionLabel === "string"
+        ? record[optionLabel] // Related model field API id
+        : optionLabel(record) // Callback on the whole related model record
+      : record.name; // TODO - Incorporate DataModel.defaultDisplayField
+
+  return (
+    records?.map((record: Record<string, any>) => ({
+      id: record.id,
+      label: getRecordLabel(record, optionLabel),
+    })) ?? []
+  );
+};
+
+/**
+ * For getting the selected record in a BelongsTo relationship
+ * Returns the selected record in an array for interoperability with the HasOne/HasMany hook
+ *
+ * The lookup is done using the `findBy` to lookup on the current model to retrieve the related model record data
+ */
+export const useLinkedChildModelRelatedModelRecords = (props: AutoRelationshipInputProps) => {
```

Comment:
One thing to think about is that all these hooks would be really useful for someone building their own has many selector component, but right now are hard-coupled to the autoform context and metadata. We don't have to adjust it right now, but I'd love to strive to make these things not need any metadata at all and instead just get given the lowest level stuff they need as input so they could eventually be exposed as headless hooks for folks who want the fine-grained UX control. I really like that we're bottling up the functionality in hooks, just wanna make em maximally reusable.

Reply:
WIP

Currently in the midst of refactoring this such that AutoForm specific stuff usage is minimized/eliminated. A few things remain that I want to come back to with fresh eyes